File: tks/ApplicationCore/ApplicationServices/src/test/kotlin/application/services/ResourcesServiceTest.java
package application.services;

import domain.exceptions.IncompatibleResourceFormatException;
import domain.exceptions.ResourceBlockedByRentException;
import domain.exceptions.ResourceNotFoundException;
import domain.exceptions.UnknownResourceException;
import application.helpers.AccessionNumberHelper;
import ports.secondary.RentSearchPort;
import lombok.SneakyThrows;
import domain.model.values.AccessionNumber;
import domain.model.values.Email;
import ports.secondary.ResourcePersistencePort;
import ports.secondary.ResourceSearchPort;
import domain.model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Date;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class ResourcesServiceTest {

    ResourcesService resourcesService;

    Resource sampleBook;
    Resource sampleMagazine;
    Resource invalidResource;
    AccessionNumber sampleAccessionNumber;

    @Mock
    ResourcePersistencePort resourcePersistencePort;
    @Mock
    ResourceSearchPort resourceSearchPort;
    @Mock
    RentSearchPort rentSearchPort;

    @BeforeEach
    public void init() {
        sampleAccessionNumber = AccessionNumberHelper.generate();
        resourcesService = new ResourcesService(resourcePersistencePort, resourceSearchPort, rentSearchPort);
        sampleBook = new Book(null, "Diuna", "Frank Herbert");
        sampleMagazine = new Magazine(null, "Świerszczyk", "Nowa era");
        invalidResource = new InvalidResource(null, "invalid");
    }

    @Test
    public void GivenResourceOfValidType_Create_ShouldSuccess() throws UnknownResourceException {
        resourcesService.create(sampleBook);
        verify(resourcePersistencePort).add(eq(sampleBook));
        assertNotNull(sampleBook.getAccessionNumber());
    }

    @Test
    public void GivenResourceOfInvalidType_Create_ShouldFail() {
        assertThrows(UnknownResourceException.class, () -> resourcesService.create(invalidResource));
        verify(resourcePersistencePort, never()).add(any());
    }

    @Test
    public void GivenResourceWithNotNullId_Create_ShouldFail() {
        sampleBook.setAccessionNumber(sampleAccessionNumber);
        assertThrows(UnknownResourceException.class, () -> resourcesService.create(sampleBook));
        verify(resourcePersistencePort, never()).add(any());
    }

    @Test
    public void GivenValidResourceWithNotNullId_Update_ShouldSuccess() {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleAccessionNumber))).thenReturn(sampleMagazine);
        resourcesService.update(sampleMagazine);
        verify(resourcePersistencePort).save(sampleMagazine);
    }

    @Test
    public void GivenNonExistingResource_Update_ShouldFail() {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleMagazine.getAccessionNumber()))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.update(sampleMagazine));
        verify(resourcePersistencePort, never()).save(any());
    }

    @Test
    public void GivenOtherResourceType_Update_ShouldFail() {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        sampleBook.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleAccessionNumber))).thenReturn(sampleMagazine);
        assertThrows(IncompatibleResourceFormatException.class, () -> resourcesService.update(sampleBook));
        verify(resourcePersistencePort, never()).save(any());
    }

    @Test
    public void GivenValidResourceId_Remove_ShouldSuccess() {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleAccessionNumber))).thenReturn(sampleMagazine);
        when(rentSearchPort.findActiveByResourceId(eq(sampleAccessionNumber))).thenReturn(null);
        resourcesService.remove(sampleAccessionNumber);
        verify(resourcePersistencePort).remove(sampleMagazine);
    }

    @Test
    public void GivenInvalidResourceId_Remove_ShouldFail() {
        when(resourceSearchPort.findById(eq(sampleAccessionNumber))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.remove(sampleAccessionNumber));
        verify(resourcePersistencePort, never()).remove(any());
    }

    @SneakyThrows
    @Test
    public void GivenRentResourceId_Remove_ShouldFail() {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleAccessionNumber))).thenReturn(sampleMagazine);
        when(rentSearchPort.findActiveByResourceId(eq(sampleAccessionNumber))).thenReturn(new Rent(UUID.randomUUID(), new Date(), null, new Email("mszewc@edu.pl"), sampleAccessionNumber));
        assertThrows(ResourceBlockedByRentException.class, () -> resourcesService.remove(sampleAccessionNumber));
        verify(resourcePersistencePort, never()).remove(any());
    }

    @Test
    public void GivenValidResourceId_GetDetails_ShouldSuccess() throws ResourceNotFoundException {
        sampleMagazine.setAccessionNumber(sampleAccessionNumber);
        when(resourceSearchPort.findById(eq(sampleMagazine.getAccessionNumber()))).thenReturn(sampleMagazine);
        var result = resourcesService.getDetails(sampleMagazine.getAccessionNumber());
        verify(resourceSearchPort).findById(eq(sampleMagazine.getAccessionNumber()));
        assertEquals(sampleMagazine, result);
    }

    @Test
    public void GivenInvalidResourceId_GetDetails_ShouldFail() {
        when(resourceSearchPort.findById(eq(ResourcesServiceTest.this.sampleAccessionNumber))).thenReturn(null);
        assertThrows(ResourceNotFoundException.class, () -> resourcesService.getDetails(ResourcesServiceTest.this.sampleAccessionNumber));
    }
}

File: tks/ApplicationCore/ApplicationServices/src/test/kotlin/application/services/ResourcesServiceTest.kt
package application.services

import application.helpers.AccessionNumberHelper
import domain.exceptions.IncompatibleResourceFormatException
import domain.exceptions.ResourceBlockedByRentException
import domain.exceptions.ResourceNotFoundException
import domain.exceptions.UnknownResourceException
import domain.model.Book
import domain.model.Magazine
import domain.model.context.rents.Rent
import domain.model.traits.Resource
import domain.model.values.AccessionNumber
import domain.model.values.Email
import io.mockk.every
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import ports.secondary.RentSearchPort
import ports.secondary.ResourcePersistencePort
import ports.secondary.ResourceSearchPort
import java.util.*

@ExtendWith(MockKExtension::class)
class ResourcesServiceTest {

    data class InvalidResource(
        override var accessionNumber: AccessionNumber?,
        override var title: String
    ) : Resource

    private lateinit var resourcesService: ResourcesService
    private lateinit var sampleBook: Resource
    private lateinit var sampleMagazine: Resource
    private lateinit var invalidResource: Resource
    private val sampleAccessionNumber: AccessionNumber = AccessionNumberHelper.generate()

    @RelaxedMockK
    lateinit var resourcePersistencePort: ResourcePersistencePort

    @RelaxedMockK
    lateinit var resourceSearchPort: ResourceSearchPort

    @RelaxedMockK
    lateinit var rentSearchPort: RentSearchPort

    @BeforeEach
    fun init() {
        resourcesService = ResourcesService(resourcePersistencePort, resourceSearchPort, rentSearchPort)
        sampleBook = Book(null, "Diuna", "Frank Herbert")
        sampleMagazine = Magazine(null, "Świerszczyk", "Nowa era")
        invalidResource = InvalidResource(null, "invalid")
    }

    @Test
    fun `Given resource of valid type then create should success`() {
        resourcesService.create(sampleBook)
        verify(exactly = 1) { (resourcePersistencePort).save(sampleBook) }
        Assertions.assertNotNull(sampleBook.accessionNumber)
    }

    @Test
    fun `Given resource of invalid type then create should fail`() {
        Assertions.assertThrows(UnknownResourceException::class.java) {
            resourcesService.create(
                invalidResource
            )
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given resource with not null id then create should fail`() {
        sampleBook.accessionNumber = sampleAccessionNumber
        Assertions.assertThrows(UnknownResourceException::class.java) {
            resourcesService.create(sampleBook)
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given valid resource with not null id then update should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { (resourceSearchPort.findByAccessionNumber(sampleAccessionNumber)) }.returns(sampleMagazine)
        resourcesService.update(sampleMagazine)
        verify(exactly = 1) { (resourcePersistencePort).save(sampleMagazine) }
    }

    @Test
    fun `Given non existing resource then update should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.update(sampleMagazine)
        }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given different resource type then update should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        sampleBook.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        Assertions.assertThrows(
            IncompatibleResourceFormatException::class.java
        ) { resourcesService.update(sampleBook) }
        verify(exactly = 0) { resourcePersistencePort.save(any()) }
    }

    @Test
    fun `Given valid resource rd then remove should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        every { rentSearchPort.findActiveByResourceId(sampleAccessionNumber) } returns null
        resourcesService.remove(sampleAccessionNumber)
        verify(exactly = 1) { resourcePersistencePort.remove(sampleMagazine) }
    }

    @Test
    fun `Given invalid resource id then remove should fail`() {
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.remove(sampleAccessionNumber)
        }
        verify(exactly = 0) { resourcePersistencePort.remove(any()) }
    }

    @Test
    fun `Given rent resource id then remove should fail`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns sampleMagazine
        every { rentSearchPort.findActiveByResourceId(sampleAccessionNumber) } returns Rent(
            UUID.randomUUID(),
            Date(),
            null,
            Email("mszewc@edu.pl"),
            sampleAccessionNumber
        )
        Assertions.assertThrows(ResourceBlockedByRentException::class.java) {
            resourcesService.remove(sampleAccessionNumber)
        }
        verify(exactly = 0) { resourcePersistencePort.remove(any()) }
    }

    @Test
    fun `Given valid resource id then getDetails should success`() {
        sampleMagazine.accessionNumber = sampleAccessionNumber
        every { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) } returns sampleMagazine

        val result = resourcesService.getDetails(sampleMagazine.accessionNumber!!)

        verify(exactly = 1) { resourceSearchPort.findByAccessionNumber(sampleMagazine.accessionNumber!!) }
        Assertions.assertEquals(sampleMagazine, result)
    }

    @Test
    fun `Given invalid resource id then getDetails should fail`() {
        every { resourceSearchPort.findByAccessionNumber(sampleAccessionNumber) } returns null
        Assertions.assertThrows(ResourceNotFoundException::class.java) {
            resourcesService.getDetails(sampleAccessionNumber)
        }
    }
}